### PR TITLE
HTML-761: Obs tag with answerClasses to filter out retired concepts.

### PR DESCRIPTION
**Pull request by sherrif10**
https://issues.openmrs.org/browse/HTML-761
cc @mks-d  @ibacher

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java [hunk 1]
```diff
@@ -698,7 +698,8 @@ private void prepareWidgets(FormEntryContext context, Map<String, String> parame
 							if (cc == null) {
 								throw new RuntimeException("Cannot find concept class " + className);
 							}
-							conceptAnswers.addAll(Context.getConceptService().getConceptsByClass(cc));
+							conceptAnswers.addAll((Collection<? extends Concept>) Context.getConceptService()
+							        .getConceptsByClass(cc).stream().filter(it -> !it.getRetired()));
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java:702]:
```suggestion
							conceptAnswers.addAll(Context.getConceptService()
							        .getConceptsByClass(cc).stream().filter(it -> !it.getRetired())
							        .collect(Collectors.toList()));
```

Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java]: Resolved thanks

**Review by mogoodrich — COMMENTED** [review]
LGTM... but wondering if @mseaton  has any thoughts on this as he recently did a (somewhat) related change to HFE.

**Comment by sherrif10** [comment]
Thanks @mogoodrich

**Review by mseaton — APPROVED** [review]
The PR that I'm working on for HTML-765 _should_ ensure that previously saved values will still appear in edit mode, so I don't think this will adversely affect that use case, assuming I get that in.

I'm not 100% sure I agree this fix is needed, but I don't see any harm in it I guess.  One could argue that if a Concept is explicitly added as a valid Answer to a Concept, then it should be displayed, and if you don't want it to be listed any more, just remove it from the allowed answers to a given Question.  @mks-d - looks like you created this ticket, so I'll defer to you on whether you've given that particular thought.

**Comment by sherrif10** [comment]
Thanks @mseaton for clarifying, So the idea was that if any concept is retired , it shouldnot reflect again. cc @mks-d

**Comment by mogoodrich** [comment]
@mseaton  I think this only applies to the "answerClasses" attribute, which displays all concepts that depend to a class or set of classes...

**Comment by sherrif10** [comment]
@mogoodrich  Yeah sure basically coded concepts with answerClasses Attribute, so the idea is that Obs tag should not return retired coded concept

**Comment by sherrif10** [comment]
Hello @mks-d  Any more suggestion on this work being done here thanks

**Comment by sherrif10** [comment]
Shall we have a merge into master because i think its ready cc @mks-d

**Review by sherrif10 — COMMENTED** [review]
LGTM  this can be merged cc @mksd @dkayiwa  thnks

**Review by mks-d — CHANGES_REQUESTED** [review]
@mseaton about your comment [here](https://github.com/openmrs/openmrs-module-htmlformentry/pull/216#pullrequestreview-543344284), as @mogoodrich said the scope here is only about the `answerClasses` attribute.

LGTM except for the **.gitignore** file that should be untouched.

File: api-2.3/.gitignore [hunk 2]
```diff
@@ -0,0 +1 @@
+/target/
```
[the remark this answers is not in the document]
Comment (sherrif10) [reply api-2.3/.gitignore:1]: ohh sure let me fix it

[the remark this answers is not in the document]
Comment (sherrif10) [reply api-2.3/.gitignore]: Resolved thanks

File: api-2.3/.gitignore [hunk 3]
```diff
@@ -0,0 +1 @@
+
```
Comment (mks-d) [inline api-2.3/.gitignore:1]: Still there.

**Review by mks-d — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java [hunk 4]
```diff
@@ -698,7 +699,8 @@ private void prepareWidgets(FormEntryContext context, Map<String, String> parame
 							if (cc == null) {
 								throw new RuntimeException("Cannot find concept class " + className);
 							}
-							conceptAnswers.addAll(Context.getConceptService().getConceptsByClass(cc));
+							conceptAnswers.addAll(Context.getConceptService().getConceptsByClass(cc).stream()
+							        .filter(it -> !it.getRetired()).collect(Collectors.toList()));
```
Comment (mks-d) [inline api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java:703]: A last detail, in `it -> !it.getRetired()` just change the variable name from `it` to `c` ("c" for "concept"). More readable.

**Review by mks-d — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java (hunk 4 above)
Comment (sherrif10) [reply api/src/main/java/org/openmrs/module/htmlformentry/element/ObsSubmissionElement.java]: Thanks @mks-d  have responded to the comment  above

**Review by mks-d — APPROVED** [review]

**Pull request merged**